Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -102,14 +121,33 @@ private async Task<ReadHttpResponseMessage> SendAsync<TRequest>(RestMethod restM
             }
         }
 
-        private async Task AddAuthHeadersAsync(HttpRequestMessage request, RestMethod method)
+        private async Task AddAuthHeadersAsync(HttpRequestMessage request, RestMethod restMethod, CancellationToken combinedCancellationToken)
         {
             Uri hostUri = request.RequestUri.IsAbsoluteUri ? request.RequestUri : _httpClient.BaseAddress;
             string schemeAndAuthority = hostUri.GetLeftPart(UriPartial.Authority);
 
             var metadata = new Metadata();
-            var context = new AuthInterceptorContext(schemeAndAuthority, method.FullName);
-            await _channelAuthInterceptor(context, metadata).ConfigureAwait(false);
+            var context = new AuthInterceptorContext(schemeAndAuthority, restMethod.FullName);
+
+            var combinedCancellationTask = Task.Delay(-1, combinedCancellationToken);
+            var channelTask = _channelAuthInterceptor(context, metadata);
+            var resultTask = await Task.WhenAny(channelTask, combinedCancellationTask).ConfigureAwait(false);
+            if (combinedCancellationTask == resultTask)
+            {
+                throw new TaskCanceledException(
+                    $"The task was cancelled by the caller while waiting for the auth headers to be added for a method `{restMethod.FullName}`.");
+            }
+
+            if (channelTask.IsFaulted)
+            {
+                if (channelTask.Exception is object)
+                {
+                    throw channelTask.Exception;
```

Comment:
I think we'd want `ExceptionDispatchInfo.Capture(channelTask.Exception).Throw();` here, to preserve the stack trace.

Reply:
👍